### ISSUE TITLE
Materials Project Training

### ISSUE THREAD

**Issue by jstoppelman** [atomistic-machine-learning/schnetpack#226]
Hello, I want to eventually train SchNet models on some crystal structures that I have obtained using Quantum ESPRESSO, and I figured I would first test out my model settings on the Materials Project database. While training the formation energy per atom, the MAE seems to plateau at 0.6 eV/atom, which is quite a bit worse than the 0.041 eV/atom that I see in the SchNetPack paper: https://doi.org/10.1021/acs.jctc.8b00908 I have the code I used to construct the model below. Are the settings I used reasonable and similar for the ones used in the benchmark, or is there something I'm missing? 
```
mp = MaterialsProject("./mp.db", load_only=[MaterialsProject.EformationPerAtom], environment_provider=sch.environment.AseEnvironmentProvider(5))

train, val, test = spk.train_test_split(
        data=mp,
        num_train=int(4/5*len(mp)),
        num_val=int(1/5*len(mp)),
        split_file="split.npz"
    )

train_loader = spk.AtomsLoader(train, batch_size=16, shuffle=True)
val_loader = spk.AtomsLoader(val, batch_size=16)

schnet = spk.representation.SchNet(
    n_atom_basis=128,
    n_filters=128,
    n_gaussians=30,
    n_interactions=6,
    cutoff=5., cutoff_network=spk.nn.cutoff.CosineCutoff
)

output = spk.atomistic.Atomwise(n_in=128, property=MaterialsProject.EformationPerAtom)
model = spk.AtomisticModel(representation=schnet, output_modules=output)
optimizer = Adam(model.parameters(), lr=0.0001)

loss = trn.build_mse_loss([MaterialsProject.EformationPerAtom])

metrics = [spk.metrics.MeanAbsoluteError(MaterialsProject.EformationPerAtom)]

hooks = [
    trn.CSVHook(log_path='./', metrics=metrics),
        trn.ReduceLROnPlateauHook(
        optimizer,
        patience=5, factor=0.8, min_lr=1e-6,
        stop_after_min=True
    )
]

trainer = trn.Trainer(
    model_path='./',
    model=model,
    hooks=hooks,
    loss_fn=loss,
    optimizer=optimizer,
    train_loader=train_loader,
    validation_loader=val_loader,
)
```

**Comment by Stefaanhess** [comment]
Hey @jstoppelman ,
I am not sure, but maybe your test set is very small? You use 4/5 of the data for training and 1/5 for validation. This means that there is not much left for testing? If the test set only contains 1 or two datapoints, different results can occur. Can you try to use e.g. num_train=0.8 and num_val=0.1 or something similar?
Stefaan

**Comment by jstoppelman** [comment]
@Stefaanhess Thank you for the reply! I adjusted num_train=0.8 and num_val=0.1, but the validation MAE output in the log.csv file still plateaus around 0.6 eV/atom during training, and the test MAE gives a similar result at 0.62 eV/atom. This is still pretty far off from what I would expect, and I still can't find anything wrong with the code used to setup the model...

**Comment by Stefaanhess** [comment]
I am not sure what has gone wrong... Could there be an issue with the units in the database? Did you use the automatic download of the MaterialsProject class to generate the database?

**Comment by mgastegger** [comment]
Did you try using a different patience/learning rate? In the code examples you use a `lr=0.0001` and `patience=5`, which might both be to low. The low patience in particular could cause the plateau you see. In the paper we used values of `lr=0.001` and `patience=25`.

**Comment by jstoppelman** [comment]
Reporting back after changing the `lr`. The MAE does improve to 0.36 eV/atom in comparison to before. Is it strange that the training set loss improves throughout the training, while the validation loss essentially plateaus? Going through the code, I believe that there are no unit conversions from the initial eV that the dataset is downloaded with using `matproj.py`, is this correct? I was trying to double check the units to make sure that they aren't changing for some reason, but it doesn't look like they do.
![mae_matproj](https://user-images.githubusercontent.com/30512460/85082884-c9140f80-b195-11ea-91bd-37e81d1110f8.png)

**Comment by Dankomaister** [comment]
> Reporting back after changing the `lr`. The MAE does improve to 0.36 eV/atom in comparison to before. Is it strange that the training set loss improves throughout the training, while the validation loss essentially plateaus? Going through the code, I believe that there are no unit conversions from the initial eV that the dataset is downloaded with using `matproj.py`, is this correct? I was trying to double check the units to make sure that they aren't changing for some reason, but it doesn't look like they do.
> ![mae_matproj](https://user-images.githubusercontent.com/30512460/85082884-c9140f80-b195-11ea-91bd-37e81d1110f8.png)

Hi!
What you have there is a clear example of overfitting! were the training loss decreases while the validation loss increases.
I have experienced similar results when training SchNet models for my current paper. Could SchNet benefit from adding dropout to the network? Is this anything that has been tested before?

**Comment by Stefaanhess** [comment]
@jstoppelman yes, the units are not converted in any way. I also checked that. Still I do not really know what is goinig on... I have not trained any MaterialsProject models myself, but the overfitting seems not right. Could you check the size of the dataset for me? Maybe the database is corrupted and some data is missing?
@Dankomaister we are working on the building blocks of SchNetPack at the moment, in order to allow different architectures (e.g. PhysNet). This will probably also include the support of dropout layers, but I have not tested it yet.

**Comment by jstoppelman** [comment]
@Stefaanhess The training that I made the figure from is using a subset of the first 63000 configurations, with 49600 for training and 12400 for validation and I left 1000 points for testing. I did this so the size of my training + validation set was the same as the SchNetPack paper. The total size of the Materials Project database is 124478 crystals currently, I only created a subset of the structures to see if that created any change in the training. Here is the command I used to initialize the database.
```
sub = [i for i in range(63000)]
mp = MaterialsProject("../../mp.db", load_only=[MaterialsProject.EformationPerAtom], subset=sub, environment_provider=sch.environment.AseEnvironmentProvider(5))
```
 Thanks for the help!

**Comment by Stefaanhess** [comment]
@jstoppelman I will run the MaterialsProject-script myself in order to check the results. Maybe that can help to narrow down the problem...

**Comment by jstoppelman** [comment]
@Stefaanhess ok, thank you for checking. I haven't found anything wrong with my script yet, although there could be something I'm still missing that I don't see.

**Comment by Manyi-Yang** [comment]
> @jstoppelman yes, the units are not converted in any way. I also checked that. Still I do not really know what is goinig on... I have not trained any MaterialsProject models myself, but the overfitting seems not right. Could you check the size of the dataset for me? Maybe the database is corrupted and some data is missing?
> @Dankomaister we are working on the building blocks of SchNetPack at the moment, in order to allow different architectures (e.g. PhysNet). This will probably also include the support of dropout layers, but I have not tested it yet.

@Stefaanhess  I would like to know why you think that the following picture is overfitting was not correct? I also have this kind of problem. the loss of the training set is much lower than the validation one. How can we solve this kind of problem? Thank you. 
![image](https://user-images.githubusercontent.com/50021753/132591144-e525afdb-e84d-4146-adea-0652f9e22321.png)

**Comment by Stefaanhess** [comment]
Hey @Manyi-Yang ,
what dataset are you training on and what settings are you using?

**Comment by Manyi-Yang** [comment]
@Stefaanhess  Thank you for your reply.
I am a new use of Schnet and was confused a lot when I start to use Schnet.  But I am sure you can help me to solve all the problems. 

My system is a single element liquid with 128 atoms,  the box size is about (10.0, 12.0 24.0), PBC is used. energy and forces are trained.  Here is my setting file, which gives me the best results among all my tests.  BTW, is there any tutorial for build this .json file?

> {
    "aggregation_mode": "sum",
    "batch_size": 4,
    "checkpoint_interval": 1,
    "contributions": null,
    "cuda": true,
    "cutoff": 8.0,
    "cutoff_function": "cosine",
    "datapath": "model/raw_dataset.db",
    "dataset": "custom",
    "derivative": "forces",
    "environment_provider": "torch",
    "features": 128,
    "force": null,
    "interactions": 6,
    "keep_n_checkpoints": 3,
    "log_every_n_epochs": 1,
    "logger": "csv",
    "lr": 0.0005,
    "lr_decay": 0.8,
    "lr_min": 1e-06,
    "lr_patience": 50,
    "max_epochs": 10000,
    "max_steps": null,
    "mode": "train",
    "model": "schnet",
    "modelpath": "model/",
    "n_epochs": 10000,
    "negative_dr": true,
    "normalize_filter": false,
    "num_gaussians": 50,
    "output_module": "atomwise",
    "overwrite": false,
    "parallel": true,
    "property": "energy",
    "rho": {},
    "seed": null,
    "split": [
        800,
        100
    ],
    "split_path": null,
    "stress": null
}

![image](https://user-images.githubusercontent.com/50021753/132672749-e98e1a94-74ee-42ea-a03f-9f3c1b739ef9.png)
![image](https://user-images.githubusercontent.com/50021753/132674171-06928b3a-626b-4fb0-9307-940d334ee315.png)

From these two pictures, I found that the loss for the validation set is much higher than the train set. In addition, the MAE force and energy are too large.  According to my test results,  I have the following questions:

1. when I prepared my custom dataset, what units for energy, forces, and coordinates should I keep? Is eV, eV/A, and A OK? In this case,  eV, eV/A, and A are used.  In the log.csv file, the MAE energy and force unit is eV/molecule (128 atoms) and eV/A, right?
2.  Since my box is about  (10.0, 12.0 24.0), is it OK to set the cutoff as 8.0, which is larger than  10.0/2 = 5.0?

Thank you very much!

**Comment by Stefaanhess** [comment]
Hey @Manyi-Yang ,
the units should be ok. The cutoff should also work, but you could also try something smaller (e.g. 5 Ang).
From your args file I would suggest to select a smaller number of `n_gaussians=25` and the number of training points looks a bit small. Does your dataset allow more training points?
We are currently working on a new version of schnetpack, which will include a new representation (PaiNN). This should work better on smaller datasets! Maybe you can wait for the new version, which will be merged to master in the next weeks. You could also check out the dev-branch and install it with `pip install -e .`. The documentation might still have some problems, but the scripts should already work well.

**Comment by Manyi-Yang** [comment]
Hey @Stefaanhess 
thank you very much!
Looking forward to your PaiNN version.  
When I try to select a small value of features(32), the result is much better. also, the setting of rho really makes sense in my case.  and I will try to add more points to my training set now. 
Thanks again

**Comment by Manyi-Yang** [comment]
> Hey @Stefaanhess
> thank you very much!
> Looking forward to your PaiNN version.
> When I try to select a small value of features(32), the result is much better. also, the setting of rho really makes sense in my case. and I will try to add more points to my training set now.
> Thanks again



> Hey @Manyi-Yang ,
> the units should be ok. The cutoff should also work, but you could also try something smaller (e.g. 5 Ang).
> From your args file I would suggest to select a smaller number of `n_gaussians=25` and the number of training points looks a bit small. Does your dataset allow more training points?
> We are currently working on a new version of schnetpack, which will include a new representation (PaiNN). This should work better on smaller datasets! Maybe you can wait for the new version, which will be merged to master in the next weeks. You could also check out the dev-branch and install it with `pip install -e .`. The documentation might still have some problems, but the scripts should already work well.

Hi @Stefaanhess 
AS you suggested that PaiNN would be a good option for the small point data set.  I am interested in it very much. But after installation, and when I run `spktrain --help` I get the following problems.  Pls show me how to fix it. Thanks 
![image](https://user-images.githubusercontent.com/50021753/133883195-fd22c82b-02d0-48cf-b55a-4334f37b8648.png)

Another question, could you give me some tips on how to start the paiNN since the documents are not ready yet.
